### DISCUSSION TITLE
how to specify numa-aware memory allocation with libfork ?

### DISCUSSION THREAD

**Discussion by Frandy** [ConorWilliams/libfork#19] [Q&A]
When scheduling task, numa-aware usually means one thread will consume tasks in its own pool, then get tasks from pool on the same numa node, and get tasks from the neighbour numa node finally. 

While there is a situation, for example, we have 8 threads on two numa nodes ( thread 0\~3 on numa node 0, and thread 4\~7 on numa node 1), and we have 16 jobs.
And each job has some data, and will visit these data when run task.  suppose thread 0\~3  running jobs 0\~7, and thread 4\~7 running jobs 8~15, so data of job 0\~7 are on numa node 0, and data of job 8\~15 on numa node 1.

So there are two questions, 
1. when to allocate data for jobs
    // lazy_pool or busy_pool with 8 threads;
    // for loop to allocate data
    for(jobId = 0; jobId < jobCount-1; ++jobId)
           co_await lf::fork\[ allocateDataMem\](jobId);
   co_await lf::call\[allocateDataMem\](fianlJobId);

what's the result for allocated data on which numa node ?

2. when to run task for jobs
   for(jobId = 0; jobId < jobCount - 1; ++jobI)
      co_await lf::fork\[  runSomeTask\](jobId);
   co_await lf::call\[ runSomeTask\](jobId);

what's the result for task running on which numa node ?

suppose all those task scheduled evenly, and very balance.

**Comment by ConorWilliams** [comment]
Libfork's `lazy_pool` and `busy_pool` schedulers will ensure workers/threads push/pop task form their own queue's until they run out of local work, then they steal work from neighbours which are topologically "close". All this is handled opaquely. The workers will be distributed and pinned across the physical cores at the construction of the thread pools.

**Reply by Frandy** [reply]
Thanks for your quick reply. I have updated my question. I have posted my text by mistake, and got your reply in seconds.

**Reply by ConorWilliams** [reply]
Ahh ok, in this case you will want to ensure the data-allocation is as close to the work as possible something like this:

```cpp

auto alloc_and_work(auto, int id) -> task<...> {
    auto data = co_await just[allocateDataMem]();
    auto result = co_await just[runSomeTask]();
    // Do something with the result...
}

auto launch_n(auto) -> task<...> {

    auto iota = std::ranges::views::iota(0, jobCount);

    co_await just[for_each](iota, alloc_and_work);
}
```

This uses the `for_each` higher-order function to do the loop a bit more efficiently. If you cannot move the allocation closer to the work then I'm afraid you'll have to depend on the OS migrating the pages but after that has happened libfork will work to minimize cross-numa stealing.

**Reply by Frandy** [reply]
Yeah, packing allocate data and run task into one can solve part of the issue. While for my application, it would create job data at the beginning, and run many tasks and iterations on those jobs.   

Migrating pages may be very slow, and we can not afford it to migrate pages for every iteration.

**Reply by ConorWilliams** [reply]
Could you explain the structure of your application a little more, I think this is solvable with the explicit scheduling API

**Reply by Frandy** [reply]
OK, it is a simulation application which mainly solve Ax=b repeatedly. 
A is N x N matrix, and b is N x 1 vector. For large scale system, N can exceed $10^7$.  

Without partition and parallel, the flow is like below:
1. construct sparse matrix pattern of A. 
2. evaluate value of A and b.
3. perform LU factor for A and got A = LU, while L is lower triangle matrix and U is upper triangle matrix.
4. then solve x by forward Uy = b and backward Lx = y.
5. repeat 2\~4

To solve the equation Ax = b in parallel, we will first partition the matrix into M bottom parts and one coupling top matrix. 
Then the solve process as below:
1.  partition the matrix into M parts, and  construct sparse matrix pattern for those M bottom parts and one top matrix. Usually we will force  bottom matrix with dimension smaller enough to hold in L2 cache for each core, around $10^4$, then we will have about 1000 bottom parts and one top matrix for very large scale system.  Here we want to distribute those M bottom parts into different threads and allocate sparse matrix on corresponding numa node.
2. we evaluate and update sparse matrix value for each bottom parts in parallel.  Here we want to distribute those M bottom parts job on corresponding numa node.
3. do partial factor/forward for those M bottom parts in parallel. similar requirements as step 2.
4. form top matrix/rhs with contribution from M bottom parts, and solve top matrix. Here there would be a new level for parallel if top matrix is large and can get benefit from parallel. 
6. get partial solution from top, and solve those M bottom parts in parallel.
7. repeat 2\~6

For openmp, we can first bind openmp threads onto specified numa node. as below
```
#pragma omp  threads(availableThreads)
for(int threadIdx = 0; threadIdx < availableThreads; ++threadIdx)
     bindThreadToCoreOrNumaNode();
```

Then we know the mapping between threads and numa node, and later we can distribute jobs according to this map.

Any sugession for this application ?

**Reply by ConorWilliams** [reply]
Hi Frandy,

I've been thinking about this, I don't think libfork provides a nice way of handling this, other than explicitly scheduling all the work (which kind of defeats the point of libfork). I will keep this in mind for V4, I'm open to any suggestions of a nice API for this?

**Reply by Frandy** [reply]
Thanks. Maybe it need one argument to pass numa hint for task.

**Reply by Frandy** [reply]
Do you know the library [OpenBLAS] (https://github.com/OpenMathLib/OpenBLAS) or [BLIS](https://github.com/flame/blis) ?  They are high performance matrix/vector library and can use different multi-thread backend, such as openmp, native thread. I wonder if they can get benefit from libfork, especially for nested parallel situation.

**Reply by ConorWilliams** [reply]
I'm familiar with OpenBLAS, first time I've seen BLIS though, I suspect they could benefit if they use nested parallelism, especially with a small grain size. Any idea how hard it is to hack together a new backend for them?

**Reply by Frandy** [reply]
I'm user of OpenBLAS/BLIS, and tried to learn the detail of dgemm implementation. If just to verify the ideal, and select one blas3 function   to try, it is not that hard. But if to add backend of parallel for all function, it will have a lot of work.

**Comment by ConorWilliams** [comment]
Just make sure when compiling libfork that CMake prints something like this:

```sh
 Found hwloc 2.10.0rc3-git in /home/cj/libfork/build/rel/vcpkg_installed/x64-linux/lib/pkgconfig/../../include:hwloc;m;pthread
```

hwloc is the library that libfork uses to detect the NUMA topology
